SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityMesosOfferScheduler.java

```diff
@@ -134,32 +162,163 @@ public SingularityMesosOfferScheduler(MesosConfiguration mesosConfiguration,
+
+    List<CachedOffer> cachedOfferList = offerCache.checkoutOffers();
+    Map<String, CachedOffer> cachedOffers = new HashMap<>();
+    for (CachedOffer cachedOffer : cachedOfferList) {
+      if (isValidOffer(cachedOffer.getOffer())) {
+        cachedOffers.put(cachedOffer.getOfferId(), cachedOffer);
+        offersToCheck.add(cachedOffer.getOffer());
+      } else if (cachedOffer.getOffer().getId() != null && cachedOffer.getOffer().getId().getValue() != null) {
+        mesosSchedulerClient.decline(Collections.singletonList(cachedOffer.getOffer().getId()));
+        offerCache.rescindOffer(cachedOffer.getOffer().getId());
+      } else {
+        LOG.warn("Offer {} was not valid, but we can't decline it because we have no offer ID!");
+      }
+    }
+
+    List<CompletableFuture<Void>> slaveCheckFutures = new ArrayList<>();
+    uncached.forEach((offer) -> slaveCheckFutures.add(runAsync(() -> checkOfferAndSlave(offer, offersToCheck))));
+    CompletableFutures.allOf(slaveCheckFutures).join();
+
+    final Set<OfferID> acceptedOffers = Sets.newHashSetWithExpectedSize(offersToCheck.size());
+
+    try {
+      Collection<SingularityOfferHolder> offerHolders = checkOffers(offersToCheck);
+
+      for (SingularityOfferHolder offerHolder : offerHolders) {
+        if (!offerHolder.getAcceptedTasks().isEmpty()) {
+          List<Offer> leftoverOffers = offerHolder.launchTasksAndGetUnusedOffers(mesosSchedulerClient);
+
+          leftoverOffers.forEach((o) -> {
+            if (cachedOffers.containsKey(o.getId().getValue())) {
+              offerCache.returnOffer(cachedOffers.remove(o.getId().getValue()));
+            } else {
+              offerCache.cacheOffer(start, o);
+            }
+          });
+
+          List<Offer> offersAcceptedFromSlave = offerHolder.getOffers();
+          offersAcceptedFromSlave.removeAll(leftoverOffers);
+          offersAcceptedFromSlave.stream()
+              .filter((offer) -> cachedOffers.containsKey(offer.getId().getValue()))
+              .map((o) -> cachedOffers.remove(o.getId().getValue()))
+              .forEach(offerCache::useOffer);
+          acceptedOffers.addAll(offersAcceptedFromSlave.stream().map(Offer::getId).collect(Collectors.toList()));
+        } else {
+          offerHolder.getOffers().forEach((o) -> {
+            if (cachedOffers.containsKey(o.getId().getValue())) {
+              offerCache.returnOffer(cachedOffers.remove(o.getId().getValue()));
+            } else {
+              offerCache.cacheOffer(start, o);
+            }
+          });
+        }
+      }
+
+      LOG.info("{} remaining offers not accounted for in offer check", cachedOffers.size());
+      cachedOffers.values().forEach(offerCache::returnOffer);
+    } catch (Throwable t) {
+      LOG.error("Received fatal error while handling offers - will decline all available offers", t);
+
+      mesosSchedulerClient.decline(offersToCheck.stream()
```

Comment:
Since `offersToCheck` is created via `Collections.synchronizedList()`, iterations over it need to be inside a `synchronized (offersToCheck) { }` block